### PR TITLE
Add inverted options and strict mode

### LINKED ISSUE #2585: Add --strict for opt-in to new features
I have a codebase which is currently compatible with all the 'extra hard' checks:
`--strict-optional --warn-unused-ignores --warn-redundant-casts --check-untyped-defs --disallow-untyped-defs --disallow-untyped-calls`

I guess in future mypy will add future parameters like this, and I'd like our CI pipeline to automatically opt in to enabling them, so this functionality isn't missed.

Therefore, I'd like a `--strict` option which enables all the above options, and support for `--no-*` to disable specific options on a case-by-case basis if we find a new option isn't desirable.

Thanks!

### PR DESCRIPTION
**Pull request by ddfisher**
Alternative to #2700 with less boilerplate.  Fixes #2585.

### CODE REVIEW

**Comment by ddfisher** [comment]
@alexjurkiewicz, how does this look to you?

File: mypy/main.py [hunk 1]
```diff
@@ -135,6 +135,23 @@ def process_options(args: List[str],
                                      fromfile_prefix_chars='@',
                                      formatter_class=AugmentedHelpFormatter)
 
+    strict_flag_names = []  # type: List[str]
+    strict_flag_assignments = []  # type: List[Tuple[str, bool]]
+
+    def add_invertable_flag(flag, *, inverse, default, dest=None, help, strict_flag=False):
```
Comment (ambv) [inline mypy/main.py]: nit: s/invertable/invertible/

Comment (ddfisher) [inline mypy/main.py]: Thanks

**Comment by alexjurkiewicz** [comment]
I like it! Does this work if I run mypy with `--strict --allow-untyped-calls`? Doesn't the logic at line 348 mean strict mode cannot be overridden?

**Comment by ddfisher** [comment]
Oops!  Fixed.

**Comment by alexjurkiewicz** [comment]
OK, that logic is much cleaner :)

Regarding the param names, I am not a huge fan of a non-standard approach to option inverses (allow -> disallow, foo -> no-foo, fast -> old, more?). I think it will make understanding any given commandline much harder for those not intimately familiar with the mypy options.

Linters generally use a symbolic approach like `--enable-checks E123,E456 --disable-checks E789,012`
GCC example from the original PR uses a common prefix of `-Wno-`

Either of those I prefer.

Beyond that, 👍  and thanks 🙌 !

**Comment by alexjurkiewicz** [comment]
Because the opt-out names aren't standard, they may need to be documented somehow. Users probably can't guess that the opposite of `--check-untyped-defs` is `--ignore-untyped-defs`.

I was going to suggest adding the inverse to the end of the argument's help text (perhaps only if it's a strict option), but this might be too verbose. No other good ideas immediately...

**Comment by gnprice** [comment]
To my eye "allow"<->"disallow" and "show"<->"hide" feel very predictable and natural, but "check"<->"ignore" much less so.

I think it may be because "allow"<->"disallow" is adding/removing an explicit negation, and "show"<->"hide" is a very common opposition in a wide range of UIs, while "check" vs. "ignore" isn't a standard pairing. Especially when looking at "ignore", there are a lot of possible opposites for that and it isn't obvious that the opposite will be "check". I don't think I'd be sure even looking at the pair "ignore" and "check", in say two different scripts or a CLI invocation that was overriding a config file, that they were exact opposites and there wasn't some third alternative.

Perhaps `--no-check-untyped-defs`, but allow/disallow and show/hide?

**Comment by ddfisher** [comment]
I agree that check/ignore and fast/old aren't obvious opposites.  I'll change both of those to use the generic `--no-` prefix.

This PR does document the inverse flags, actually -- it adds them to the help output of the standard flag.

**Comment by ddfisher** [comment]
@alexjurkiewicz does this seems reasonable to you?  If so, I'll merge.

**Comment by ddfisher** [comment]
Ah, my understanding is that you're on vacation.  We'd like to have the inverted options, so I'm going to merge this (and my understanding is that the `--strict` flag is in pretty good shape here).

This PR hasn't received detailed review, but enough people have looked at it (and it's an isolated-enough part of the code) that the merge feels reasonable.

**Comment by alexjurkiewicz** [comment]
Yes, I am away, my email reply went AWOL apparently. Thanks for helping out / rewriting this feature! I really appreciate the assistance :+1:

**Comment by ddfisher** [comment]
Glad I could help!

**Pull request merged**